### ISSUE TITLE
'events' plugin design

### ISSUE THREAD

**Issue by wmhilton** [isomorphic-git/isomorphic-git#493]
# Motivation
I've been trying to figure out how best to eliminate the `emitter` parameter from `git.fetch` and `git.push`. I dislike it because it's not a pure JSON parameter, which means it can't easily be transferred via `postMessage` to an instance of isomorphic-git in a WebWorker. However, making it a global plugin instead of a per-function parameter creates a challenge of how to distinguish between the events of one function from the events of another. (One such scenario would be if the user wanted to do multiple `push` commands concurrently and show separate progress bars for each `push`; say to push all of a repo's submodules simultaneously, or to push a single branch to multiple peers simultaneously over a P2P network.)

After giving some thought as to how to tackle the problem of distinguishing events from concurrent commands (I could easily assign running git command instances unique ids, but how would I communicate these ids to the user without breaking the current "await until command is done" behavior?) I've decided I can punt the issue for now. Instead of eliminating the `emitter` argument completely, we can _substitute_ a new string argument called `eventPrefix`. So generating unique ids would be the responsibility of the user for now, which is a burden but not a horrible one.

Here is some example code:

**Comment by wmhilton** [comment]
What do you think @jcubic and @mizchi ?

**Comment by jcubic** [comment]
I don't think it will help in my, I'm creating same event emiter for each command but the emitter use reference that cannot be use in worker so I will need to have a proxy with this new plugin.

Will emitter parameter be removed or it will still work when this plugin be implemented?

**Comment by wmhilton** [comment]
> Will emitter parameter be removed or it will still work when this plugin be implemented?

It'll still work the same! Just like the `fs` parameter overrides the plugin-provided `fs`, an `emitter` parameter would override the plugin-provided `emitter`. But I do think the new `emitter` plugin system would greatly simplify your current code. Here's how:

The main advantage of doing this is you won't have to treat the `emitter` argument as a special case [here](https://github.com/jcubic/git/blob/gh-pages/js/main.js#L131-L139) and [here](https://github.com/jcubic/git/blob/gh-pages/js/git-worker.js#L101-L106). In fact, since you don't have multiple commands running concurrently, you would be able to leave `emitterPrefix` blank (it'll default to `''`), and simply forward all `'message'` events from the worker to the main thread, where you'd call `term.echo(message);`. You won't need to use the [`messageEmitter()`](https://github.com/jcubic/git/blob/gh-pages/js/main.js#L202-L220) factory function at all, because you'll only need to set up a single EventEmitter at the beginning.

**Comment by wmhilton** [comment]
I'm also planning to add a new event called `'rawMessage'` that doesn't trim the `\r` and `\n` off the ends of messages. That should make it easier to duplicate the exact look and feel of the git command line.
